File: Sources/ArgentUtilsCore/Review.swift
import Foundation

/// One review-depth level, hydrated from the shared `core/review.json`.
public struct ReviewDepth: Identifiable {
    public let id: String
    public let title: String
    public let blurb: String
    public let fragment: String
    /// The "fix it on the branch" disposition for this depth, appended only when
    /// we may actually commit. `nil` for flag-only depths / never used review-only.
    public let onBranch: String?

    public init(id: String, title: String, blurb: String, fragment: String, onBranch: String? = nil) {
        self.id = id
        self.title = title
        self.blurb = blurb
        self.fragment = fragment
        self.onBranch = onBranch
    }
}

/// Read-only access to the review-prompt model in `core/review.json`.
public enum ReviewCatalog {
    public static func depths() -> [ReviewDepth] {
        guard let r = try? CoreAssets.review() else { return [] }
        return r.depths.map { ReviewDepth(id: $0.id, title: $0.title, blurb: $0.blurb, fragment: $0.fragment, onBranch: $0.onBranch) }
    }
    public static func defaultDepthID() -> String {
        (try? CoreAssets.review())?.defaultDepth ?? depths().first?.id ?? ""
    }
    public static func depth(id: String) -> ReviewDepth {
        let all = depths()
        if let match = all.first(where: { $0.id == id }) { return match }
        if let def = all.first(where: { $0.id == defaultDepthID() }) { return def }
        return all.first ?? ReviewDepth(id: "", title: "", blurb: "", fragment: "")
    }
}

/// Who authored a specific PR under review, when known. Selects the prompt (fix-on-
/// branch vs review-only vs author-gated) and which action toggles even apply.
public enum SpecificAuthor: Equatable {
    case unknown   // specific PR, author not polled yet / poll failed — offer everything
    case mine      // fix on the branch (CASE A)
    case theirs    // review only (CASE B)
}

/// Everything the Review-PRs wizard collects, plus the logic that turns it into
/// the prompt handed to a fresh `claude` session. Pure value type — the prompt
/// text comes from `core/review.json`; only the assembly order/conditions live
/// here, shared verbatim with the Linux front-end.
public struct ReviewConfig {
    /// Whose PRs we review — the same axis the Resolve-conflicts wizard uses.
    public typealias Target = PRTarget

    public var depth: String          // depth id; "" -> default
    public var target: Target
    public var username: String
    /// The authenticated viewer login (from the Store), used as the @handle for "mine".
    public var me: String

    public var markReady: Bool        // mark perfectly-clean PRs ready for review
    public var leaveReviews: Bool     // formal review — for OTHERS' PRs (and a specific PR)
    public var replyToReviews: Bool   // reply to others' threads — on MY PRs (and a specific PR)

    public var includeDrafts: Bool
    public var includeReady: Bool
    public var specificPR: String

    /// The "final pass" escalation: a culminating full-E2E verdict pass. Off by default.
    public var finalPass: Bool

    /// For a specific PR: whether it's mine, someone else's, or not yet determined. The
    /// wizard polls the PR's author and sets this; the monitors set it directly (they
    /// always know). Ignored unless single-PR.
    public var specificAuthor: SpecificAuthor

    public init(depth: String = "", target: Target = .mine, username: String = "",
                me: String = "", markReady: Bool = true, leaveReviews: Bool = true,
                replyToReviews: Bool = true, includeDrafts: Bool = true,
                includeReady: Bool = true, specificPR: String = "", finalPass: Bool = false,
                specificAuthor: SpecificAuthor = .unknown) {
        self.depth = depth.isEmpty ? ReviewCatalog.defaultDepthID() : depth
        self.target = target
        self.username = username
        self.me = me
        self.markReady = markReady
        self.leaveReviews = leaveReviews
        self.replyToReviews = replyToReviews
        self.includeDrafts = includeDrafts
        self.includeReady = includeReady
        self.specificPR = specificPR
        self.finalPass = finalPass
        self.specificAuthor = specificAuthor
    }

    /// The @handle whose PRs we go through (empty in single-PR mode).
    public var authorHandle: String {
        switch target {
        case .mine:
            return me.isEmpty ? "me" : me
        case .someone:
            let u = username.trimmingCharacters(in: .whitespaces)
            return u.isEmpty ? "" : u
        case .specific:
            return ""
        }
    }

    /// The review disposition: mine (fix on branch) or theirs (review only). For a
    /// whose-PRs sweep it follows the target; for a specific PR it's the polled author.
    /// `.unknown` (specific, author still pending) offers every toggle, gated prompt.
    public var disposition: SpecificAuthor {
        switch target {
        case .mine: return .mine
        case .someone: return .theirs
        case .specific: return specificAuthor
        }
    }
    // Which action toggles apply. Mine-only toggles (mark-ready, reply-to-threads) hide
    // for theirs; theirs-only toggles (formal review, final verdict) hide for mine.
    // `.unknown` (author pending) leaves all four visible.
    public var canMarkReady: Bool { disposition != .theirs }
    public var canLeaveReviews: Bool { disposition != .mine }
    public var canReplyToReviews: Bool { disposition != .theirs }
    public var canFinalPass: Bool { disposition != .mine }
    public var effMarkReady: Bool { markReady && canMarkReady }
    public var effLeaveReviews: Bool { leaveReviews && canLeaveReviews }
    public var effReplyToReviews: Bool { replyToReviews && canReplyToReviews }
    public var effFinalPass: Bool { finalPass && canFinalPass }

    /// Review exactly one PR by number/URL instead of a whose-PRs sweep.
    public var isSinglePR: Bool { target == .specific }

    /// Reviewing someone else's PRs: a hard look-don't-touch mode. We never
    /// commit or push to their branch, so the no-commit guard goes in and the
    /// commit-authoring guidance stays out. (A specific PR may be mine, so it is
    /// deliberately NOT review-only.)
    public var isReviewOnly: Bool { target == .someone }

    /// The configured target repo (owner, repo), from the shared core config.
    public var targetRepo: (owner: String, repo: String) {
        let cfg = try? CoreAssets.config()
        return (cfg?.owner ?? "software-mansion", cfg?.repo ?? "argent")
    }

    /// The single-PR field parsed as a number / URL / `owner/repo#n` shorthand,
    /// checked against the target repo.
    public var prRef: PRRef {
        let (owner, repo) = targetRepo
        return PRRef.parse(specificPR, owner: owner, repo: repo)
    }

    public var isValid: Bool {
        if isSinglePR { return prRef.isValid }
        // A whose-PRs sweep needs a handle and at least one PR-state box ticked.
        return !authorHandle.isEmpty && (includeDrafts || includeReady)
    }

    private func prKind(_ scope: [String: String]) -> String {
        switch (includeDrafts, includeReady) {
        case (true, true):  return scope["prKindBoth"] ?? ""
        case (true, false): return scope["prKindDrafts"] ?? ""
        default:            return scope["prKindReady"] ?? ""
        }
    }

    public func buildPrompt() -> String {
        let review = try? CoreAssets.review()
        let scope = review?.scope ?? [:]
        let blocks = review?.blocks ?? [:]
        let cfg = try? CoreAssets.config()
        let owner = cfg?.owner ?? "software-mansion"
        let repo = cfg?.repo ?? "argent"

        // A specific PR may be mine OR someone else's — which decides whether the agent
        // may touch the branch. When we know the author (the wizard polled it, or a
        // monitor set it) emit the mine-only / review-only prompt directly; otherwise
        // (author still pending) fall back to the author-gated CASE A/B prompt.
        if isSinglePR {
            switch specificAuthor {
            case .mine:
                return buildKnownMinePrompt(review: review, scope: scope, blocks: blocks, owner: owner, repo: repo)
            case .theirs:
                return buildKnownTheirsPrompt(review: review, scope: scope, blocks: blocks, owner: owner, repo: repo)
            case .unknown:
                return buildSpecificPrompt(review: review, scope: scope, blocks: blocks, owner: owner, repo: repo)
            }
        }

        var out: [String] = []

        let tmpl = target == .mine ? (scope["scopeMine"] ?? "") : (scope["scopeOther"] ?? "")
        let scopeText = tmpl
            .replacingOccurrences(of: "{prKind}", with: prKind(scope))
            .replacingOccurrences(of: "{handle}", with: authorHandle)
        out.append((scope["multi"] ?? "")
            .replacingOccurrences(of: "{scope}", with: scopeText)
            .replacingOccurrences(of: "{owner}", with: owner)
            .replacingOccurrences(of: "{repo}", with: repo))

        // For someone else's PRs, frame the whole task as look-don't-touch up front.
        if isReviewOnly, let b = blocks["reviewOnly"] { out.append(b) }
        let chosen = ReviewCatalog.depth(id: depth)
        out.append(chosen.fragment)
        // The depth's on-branch fix step only when we may actually commit — never
        // for someone else's branch, which we don't touch.
        if !isReviewOnly, let ob = chosen.onBranch, !ob.isEmpty { out.append(ob) }
        if let bar = blocks["bar"] { out.append(bar) }
        if effMarkReady, let b = blocks["markReady"] { out.append(b) }
        if effLeaveReviews, let b = blocks["leaveReviews"] { out.append(b) }
        if effReplyToReviews, let b = blocks["reply"] { out.append(b) }
        if let trailer = blocks["trailer"] { out.append(trailer) }
        // Commit-authoring guidance only when we might actually commit — never
        // for someone else's branch, which we don't touch.
        if !isReviewOnly, let b = blocks["noAttribution"] { out.append(b) }
        if effFinalPass, let b = blocks["finalPass"] { out.append(b) }

        return out.joined(separator: "\n\n")
    }

    /// The single-PR prompt for a PR we ALREADY know is mine: no author poll, no
    /// CASE A/B — just the mine-only disposition (review approach + bar + fix-on-branch,
    /// with reply/mark-ready gated by their toggles). Used by the auto-fix monitor.
    private func buildKnownMinePrompt(review: CoreAssets.Review?, scope: [String: String],
                                      blocks: [String: String], owner: String, repo: String) -> String {
        let specific = review?.specific ?? [:]
        let chosen = ReviewCatalog.depth(id: depth)
        let pr = prRef.numberString
        let handle = me.isEmpty ? "me" : me
        func fill(_ s: String) -> String {
            s.replacingOccurrences(of: "{pr}", with: pr)
                .replacingOccurrences(of: "{owner}", with: owner)
                .replacingOccurrences(of: "{repo}", with: repo)
                .replacingOccurrences(of: "{me}", with: handle)
        }
        var out: [String] = []
        out.append(fill(scope["single"] ?? ""))
        out.append(fill(specific["mineOnly"] ?? ""))
        // First and foremost: screen + verify + fix/dismiss + respond to every reviewer
        // finding already on the PR, BEFORE the agent's own review.
        out.append(fill(specific["reviewerFindingsFirst"] ?? ""))
        out.append(chosen.fragment)
        if let bar = blocks["bar"] { out.append(bar) }
        if let ob = chosen.onBranch, !ob.isEmpty { out.append(ob) }
        if effMarkReady, let b = blocks["markReady"] { out.append(b) }
        if let b = blocks["noAttribution"] { out.append(b) }
        if let trailer = blocks["trailer"] { out.append(trailer) }
        // No final verdict for my own PR — I don't approve my own work. The reviewer-
        // findings block above already covers replying to threads (so the separate reply
        // block is redundant here).
        return out.filter { !$0.isEmpty }.joined(separator: "\n\n")
    }

    /// The single-PR prompt for a PR we ALREADY know is someone else's (a review
    /// requested from me): no author poll, no CASE A/B — the review-only, hands-off
    /// disposition directly (leave a formal review if the toggle is on, never touch the
    /// branch). Used by the review-request monitor.
    private func buildKnownTheirsPrompt(review: CoreAssets.Review?, scope: [String: String],
                                        blocks: [String: String], owner: String, repo: String) -> String {
        let specific = review?.specific ?? [:]
        let chosen = ReviewCatalog.depth(id: depth)
        let pr = prRef.numberString
        let handle = me.isEmpty ? "me" : me
        func fill(_ s: String) -> String {
            s.replacingOccurrences(of: "{pr}", with: pr)
                .replacingOccurrences(of: "{owner}", with: owner)
                .replacingOccurrences(of: "{repo}", with: repo)
                .replacingOccurrences(of: "{me}", with: handle)
        }
        var out: [String] = []
        out.append(fill(scope["single"] ?? ""))
        out.append(fill(specific["theirsOnly"] ?? ""))
        out.append(chosen.fragment)
        if let bar = blocks["bar"] { out.append(bar) }
        if let b = blocks["reviewOnly"] { out.append(b) }
        if effLeaveReviews, let b = blocks["leaveReviews"] { out.append(b) }
        out.append(fill(specific["otherNoMarkReady"] ?? ""))
        // Deliver the approve/changes-requested verdict only when the Final-E2E toggle
        // is on (manual review). Automatic runs leave it off → no verdict, the final
        // call stays with me.
        if effFinalPass, let b = blocks["finalPass"] {
            out.append(b)
        } else if let b = blocks["noVerdict"] {
            out.append(fill(b))
        }
        if let trailer = blocks["trailer"] { out.append(trailer) }
        return out.filter { !$0.isEmpty }.joined(separator: "\n\n")
    }

    /// The single-PR (Specific PR) prompt. Because the PR may be mine or someone
    /// else's — and that decides whether the branch may be touched — this prompt
    /// tells the agent to poll the author first, then split into two mutually
    /// exclusive cases: CASE A (mine → fix on the branch, mark clean ready, reply,
    /// no AI attribution) and CASE B (theirs → review only, never touch the branch,
    /// leave a formal review, and explicitly DO NOT mark it ready). The action
    /// sub-blocks are gated by the same toggles the wizard exposes.
    private func buildSpecificPrompt(review: CoreAssets.Review?, scope: [String: String],
                                     blocks: [String: String], owner: String, repo: String) -> String {
        let specific = review?.specific ?? [:]
        let chosen = ReviewCatalog.depth(id: depth)
        let pr = prRef.numberString
        let handle = me.isEmpty ? "me" : me
        func fill(_ s: String) -> String {
            s.replacingOccurrences(of: "{pr}", with: pr)
                .replacingOccurrences(of: "{owner}", with: owner)
                .replacingOccurrences(of: "{repo}", with: repo)
                .replacingOccurrences(of: "{me}", with: handle)
        }

        var out: [String] = []
        out.append(fill(scope["single"] ?? ""))
        out.append(fill(specific["determineAuthor"] ?? ""))
        // The review work itself is the same regardless of author; only the
        // disposition differs, so the approach + bar come before the split.
        out.append(chosen.fragment)
        if let bar = blocks["bar"] { out.append(bar) }

        // CASE A — it's mine: fix it on the branch.
        out.append(fill(specific["mineHeader"] ?? ""))
        if let ob = chosen.onBranch, !ob.isEmpty { out.append(ob) }
        if effMarkReady, let b = blocks["markReady"] { out.append(b) }
        if effReplyToReviews, let b = blocks["reply"] { out.append(b) }
        if let b = blocks["noAttribution"] { out.append(b) }

        // CASE B — it's someone else's: review only, hands off the branch.
        out.append(fill(specific["otherHeader"] ?? ""))
        if let b = blocks["reviewOnly"] { out.append(b) }
        if effLeaveReviews, let b = blocks["leaveReviews"] { out.append(b) }
        out.append(fill(specific["otherNoMarkReady"] ?? ""))

        if let trailer = blocks["trailer"] { out.append(trailer) }
        if effFinalPass, let b = blocks["finalPass"] { out.append(b) }

        return out.filter { !$0.isEmpty }.joined(separator: "\n\n")
    }
}

/// Decides whether an auto-dispatched review of a review-requested PR may carry the
/// "final pass + verdict" escalation, or must stay comments-only. Pure & data-light so
/// it's unit-testable and shared verbatim with any front-end. Each flag independently
/// withholds the verdict for one class of PR; every flag defaults ON, matching the
/// intended policy: SKILL / installer / community PRs get comments only, all else a verdict.
public struct VerdictPolicy: Equatable {
    /// Withhold the verdict when the PR changes a SKILL file.
    public var withholdOnSkill: Bool
    /// Withhold the verdict when the PR changes the installer/CLI.
    public var withholdOnInstaller: Bool
    /// Withhold the verdict when the PR's author is outside the org (a community PR).
    public var withholdOnCommunity: Bool

    public init(withholdOnSkill: Bool = true, withholdOnInstaller: Bool = true,
                withholdOnCommunity: Bool = true) {
        self.withholdOnSkill = withholdOnSkill
        self.withholdOnInstaller = withholdOnInstaller
        self.withholdOnCommunity = withholdOnCommunity
    }

    /// Author associations trusted enough for an auto-verdict; anything else is "community".
    /// Matches the long-standing gate (org members, maintainers, established contributors).
    public static let trustedAssociations: Set<String> =
        ["OWNER", "MEMBER", "COLLABORATOR", "CONTRIBUTOR"]

    public static func isCommunity(_ authorAssociation: String) -> Bool {
        !trustedAssociations.contains(authorAssociation.uppercased())
    }

    /// Human-readable reasons the verdict is withheld for this PR under this policy.
    /// Empty ⇒ the verdict is allowed.
    public func withholdReasons(files: [String], authorAssociation: String) -> [String] {
        var reasons: [String] = []
        if withholdOnSkill, files.contains(where: Filters.isSkillFile) {
            reasons.append("touches a SKILL")
        }
        if withholdOnInstaller, files.contains(where: Filters.isInstallerFile) {
            reasons.append("touches the installer")
        }
        if withholdOnCommunity, VerdictPolicy.isCommunity(authorAssociation) {
            reasons.append("community PR")
        }
        return reasons
    }

    /// The final-pass verdict is allowed only when no enabled suppressor matches.
    public func allowsVerdict(files: [String], authorAssociation: String) -> Bool {
        withholdReasons(files: files, authorAssociation: authorAssociation).isEmpty
    }
}

File: packages/diplomat-platform/linux/diplomat_app/review.py
"""Review-PRs config + prompt builder, and the Linux terminal spawner.

The prompt text (depth fragments, scope templates, action blocks) all comes from
the shared ``assets/review.json``; the *assembly* order/conditions live in Swift
(DiplomatCore/Review.swift) and are reached by shelling out to the
``diplomat-core`` CLI, so the two front-ends can't drift. ``ReviewConfig`` mirrors
the Swift struct's inputs and derived toggles, including the specific-PR author
disposition (mine / theirs / unknown), which the wizard resolves via ``gh``.

The terminal spawner is the Linux analogue of the macOS AppleScript/iTerm path:
it opens a new terminal-emulator window running ``claude "<prompt>"`` detached
from the applet.
"""

from __future__ import annotations

import json
import os
import shlex
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from enum import Enum

from . import appconfig, core
from .configbase import PRSweepConfig
from .prtarget import PRTarget


# MARK: - Specific-PR author disposition


class SpecificAuthor(Enum):
    """Who authored a specific PR under review, when known (mirrors the Swift
    ``SpecificAuthor`` enum in DiplomatCore/Review.swift). Selects the prompt
    (fix-on-branch vs review-only vs author-gated) and which action toggles apply.
    """

    UNKNOWN = "unknown"  # specific PR, author not polled yet / poll failed - offer everything
    MINE = "mine"        # fix on the branch (CASE A)
    THEIRS = "theirs"    # review only (CASE B)


def fetch_specific_author(owner: str, repo: str, number: int) -> str | None:
    """One ``gh pr view ... --json author`` -> the author login, or ``None`` on
    failure. Mirrors ``ReviewWizardView.fetchAuthor`` in ReviewWizard.swift. Runs
    the gh shell-out synchronously; call it OFF the UI thread (the wizard does).
    """
    from . import gh

    try:
        data = gh.run(
            ["pr", "view", str(number), "--repo", f"{owner}/{repo}", "--json", "author"]
        )
    except Exception:  # noqa: BLE001 - best-effort author resolution, None on any failure
        return None
    try:
        author = json.loads(data).get("author") or {}
        login = author.get("login")
        return login or None
    except (ValueError, AttributeError):
        return None


# MARK: - Review depth


def depths() -> list[dict]:
    return core.review()["depths"]


def depth_ids() -> list[str]:
    return [d["id"] for d in depths()]


def depth_by_id(depth_id: str) -> dict:
    for d in depths():
        if d["id"] == depth_id:
            return d
    # Fall back to the configured default, then the first level.
    default = core.review().get("defaultDepth")
    for d in depths():
        if d["id"] == default:
            return d
    return depths()[0]


def default_depth_id() -> str:
    return core.review().get("defaultDepth", depth_ids()[0])


# MARK: - Review config + prompt builder


@dataclass
class ReviewConfig(PRSweepConfig):
    depth: str = ""  # depth id; "" -> default
    target: PRTarget = PRTarget.MINE
    username: str = ""
    me: str = ""  # authenticated viewer login, used as the @handle for "mine"

    mark_ready: bool = True
    leave_reviews: bool = True
    reply_to_reviews: bool = True

    include_drafts: bool = True
    include_ready: bool = True
    specific_pr: str = ""
    # The "final pass" escalation: a culminating full-E2E verdict pass. Off by default.
    final_pass: bool = False
    # Soft-approve: when a review-only PR comes back perfectly clean, leave a friendly
    # "ran the sweep, all clean, thanks for contributing" comment - but NEVER an APPROVE
    # action. On by default. Outranked by final_pass (a real verdict) when both are set.
    soft_approve: bool = True

    # For a specific PR: whether it's mine, someone else's, or not yet determined.
    # The wizard polls the PR's author and sets this. Ignored unless single-PR.
    specific_author: SpecificAuthor = SpecificAuthor.UNKNOWN

    def __post_init__(self) -> None:
        if not self.depth:
            self.depth = default_depth_id()

    # author_handle / is_single_pr / target_repo / pr_ref are inherited verbatim
    # from PRSweepConfig (shared with ConflictConfig).

    # The review disposition: mine (fix on branch) or theirs (review only). For a
    # whose-PRs sweep it follows the target; for a specific PR it's the polled author
    # (UNKNOWN while pending - offers every toggle, gated prompt). Mirrors Swift's
    # ReviewConfig.disposition.
    @property
    def disposition(self) -> SpecificAuthor:
        if self.target == PRTarget.MINE:
            return SpecificAuthor.MINE
        if self.target == PRTarget.SOMEONE:
            return SpecificAuthor.THEIRS
        return self.specific_author

    # Which action toggles apply. Mine-only toggles (mark-ready, reply-to-threads)
    # hide for theirs; theirs-only toggles (formal review, final verdict) hide for
    # mine. UNKNOWN (author pending) leaves all four visible. Mirrors the Swift
    # disposition-based gates verbatim.
    @property
    def can_mark_ready(self) -> bool:
        return self.disposition != SpecificAuthor.THEIRS

    @property
    def can_leave_reviews(self) -> bool:
        return self.disposition != SpecificAuthor.MINE

    @property
    def can_reply_to_reviews(self) -> bool:
        return self.disposition != SpecificAuthor.THEIRS

    # The final approve/changes-requested verdict is a reviewer's call, so it never
    # applies to my own PRs (Swift: canFinalPass = disposition != .mine).
    @property
    def can_final_pass(self) -> bool:
        return self.disposition != SpecificAuthor.MINE

    # Soft-approve is a reviewer's courtesy on someone else's PR - never my own work
    # (Swift: canSoftApprove = disposition != .mine).
    @property
    def can_soft_approve(self) -> bool:
        return self.disposition != SpecificAuthor.MINE

    @property
    def is_valid(self) -> bool:
        if self.is_single_pr:
            return self.pr_ref.is_valid
        # A whose-PRs sweep needs a handle and at least one PR-state box ticked.
        return bool(self.author_handle) and (self.include_drafts or self.include_ready)

    def build_prompt(self) -> str:
        # Single-sourced in Swift (DiplomatCore) — assembled by the diplomat-core
        # CLI so the Linux applet can't drift from the macOS builder.
        from . import promptcore

        return promptcore.build_prompt({
            "kind": "review",
            "depth": self.depth,
            "target": self.target.name.lower(),
            "username": self.username,
            "me": self.me,
            "markReady": self.mark_ready,
            "leaveReviews": self.leave_reviews,
            "replyToReviews": self.reply_to_reviews,
            "includeDrafts": self.include_drafts,
            "includeReady": self.include_ready,
            "specificPR": self.specific_pr,
            "finalPass": self.final_pass,
            "softApprove": self.soft_approve,
            "specificAuthor": self.disposition.value,
        })


# MARK: - Terminal choice + spawning


@dataclass(frozen=True)
class SpawnTerminal:
    key: str
    title: str
    exec_name: str
    # argv inserted between the executable and `bash -c <cmd>`.
    prefix: tuple[str, ...]

    @property
    def is_installed(self) -> bool:
        return shutil.which(self.exec_name) is not None


# Ordered by preference. x-terminal-emulator (the Debian alternatives symlink) and
# the XFCE native terminal come first; xterm is the always-there fallback.
TERMINALS: list[SpawnTerminal] = [
    SpawnTerminal("x-terminal-emulator", "System default", "x-terminal-emulator", ("-e",)),
    SpawnTerminal("xfce4-terminal", "XFCE Terminal", "xfce4-terminal", ("-x",)),
    SpawnTerminal("gnome-terminal", "GNOME Terminal", "gnome-terminal", ("--",)),
    SpawnTerminal("konsole", "Konsole", "konsole", ("-e",)),
    SpawnTerminal("kitty", "kitty", "kitty", ()),
    SpawnTerminal("alacritty", "Alacritty", "alacritty", ("-e",)),
    SpawnTerminal("xterm", "xterm", "xterm", ("-e",)),
]


def terminal_by_key(key: str) -> SpawnTerminal | None:
    return next((t for t in TERMINALS if t.key == key), None)


def installed_terminals() -> list[SpawnTerminal]:
    return [t for t in TERMINALS if t.is_installed]


def default_terminal() -> SpawnTerminal:
    found = installed_terminals()
    return found[0] if found else TERMINALS[-1]  # xterm fallback


def resolved(preferred: SpawnTerminal | None) -> SpawnTerminal:
    """The terminal to actually drive: the preferred one if installed, else the
    first installed alternative, else xterm."""
    if preferred and preferred.is_installed:
        return preferred
    return default_terminal()


def default_repo_path() -> str:
    """``~/dev/<repo>`` for whichever repo the shared core config targets, so the
    fallback follows a retargeted ``assets/config.json`` instead of naming one repo."""
    return os.path.expanduser(f"~/dev/{core.config()['repo']}")


def stored_repo_path() -> str:
    """The repo root picked in Settings, or "" when unset.

    Read from the shared :mod:`appconfig` file rather than this front-end's QSettings:
    a mesh node spawns agents from its own stdlib-only process, which has no Store and
    no Qt to ask. Re-read per call, so changing the setting reaches a running node."""
    return appconfig.get(appconfig.REPO_ROOT).strip()


def repo_path() -> str:
    """The local checkout the agent works in — the ``cd`` in every spawned session.

    Strongest first: the ``DIPLOMAT_REPO`` env override (the escape hatch every other
    ``DIPLOMAT_*`` knob follows), the repo root picked in Settings, then
    :func:`default_repo_path`.
    """
    chosen = os.environ.get("DIPLOMAT_REPO") or stored_repo_path()
    return os.path.expanduser(chosen) if chosen else default_repo_path()


class SpawnError(RuntimeError):
    pass


def popen_detached(target: list[str] | str, *, shell: bool = False,
                   env: dict | None = None) -> None:
    """Launch an agent process and forget it, raising ``OSError`` if it won't start.

    Every launcher in the applet wants the same two properties, and both are
    load-bearing:

    * **its own session** (``start_new_session``) — the agent outlives the applet
      that spawned it. Without it the child shares the process group and dies with
      the tray (or with the mesh node), killing a running review mid-flight;
    * **no inherited stdio** — the applet's stdin/stdout may be a closed pipe, a
      tty it no longer owns, or the journal. A child writing into it blocks on a
      full pipe or scribbles over the parent's own output.

    ``target`` is an argv list, or a shell string with ``shell=True``. ``env``
    replaces the inherited environment (the confined foreign runner passes a
    credential-scrubbed one). The ``OSError`` is left for the caller to translate
    into its own error type, which is the only part that differs between them.
    """
    subprocess.Popen(  # noqa: S603 - argv list, or the operator's own template
        target,
        shell=shell,
        env=env,
        start_new_session=True,
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def write_prompt(prompt: str) -> str:
    # 0600 via mkstemp: /tmp is world-readable and multi-user, and a mesh
    # dispatch stages the prompt here too — don't leave it readable to other
    # local users (nor world-readable by umask).
    try:
        fd, path = tempfile.mkstemp(prefix="diplomat-review-", suffix=".txt")
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(prompt)
    except OSError as exc:
        raise SpawnError(f"Couldn't stage prompt: {exc}") from exc
    return path


def user_shell() -> str:
    """The user's interactive login shell — so the spawned command sees the aliases
    and env exported from their rc (e.g. a `claude` alias in ~/.zshrc). Override with
    DIPLOMAT_SHELL; falls back to $SHELL, then bash."""
    return os.environ.get("DIPLOMAT_SHELL") or os.environ.get("SHELL") or "/bin/bash"


def shell_command(prompt_file: str, done_path: str | None = None) -> str:
    """``cd '<repo>' 2>/dev/null; claude "$(cat '<file>')"; [printf %s $? > done;] exec "$SHELL" -i``

    Run (via :func:`user_shell`, interactively) so the user's rc is sourced and
    `claude` resolves to their alias. The trailing ``exec`` keeps the window open in
    the user's shell after the session ends.

    When ``done_path`` is given, the agent's exit code is written there the moment
    ``claude`` returns — an existence-based completion sentinel the PR auto-fix
    monitor polls to tell a still-running agent from a finished one (the Linux
    analogue of the macOS TrackedProcess done-file).
    """
    repo = shlex.quote(repo_path())
    pf = shlex.quote(prompt_file)
    done = f"printf %s $? > {shlex.quote(done_path)}; " if done_path else ""
    return f'cd {repo} 2>/dev/null; claude "$(cat {pf})"; {done}exec "$SHELL" -i'


def agent_argv(prompt_file: str, done_path: str | None = None) -> list[str]:
    """What the terminal is asked to run: the agent under the user's INTERACTIVE
    shell (``-i``, so their rc is sourced and a `claude` alias resolves — a plain
    `bash -c` gets neither), inside a tmux session of its own wherever tmux exists.

    The interactive shell is also what lets an rc take the process over before it
    reaches the ``-c`` command. The widespread "start every terminal inside tmux"
    snippet ends in ``exec tmux new-session``, guarded on ``[[ -z "$TMUX" ]]``: under
    a plain ``$SHELL -i -c``, that exec replaces the shell, the agent never runs, and
    the window shows an empty session — a spawn that reports success and launched
    nothing. Opening the session ourselves satisfies that guard, so such an rc runs
    on to the aliases instead of handing off.

    It is also the only way :mod:`tmuxwatch` can reach an agent: ``capture-pane`` and
    ``send-keys`` are what the Claude-API-error watcher reads and types through on
    Linux, and they only see panes. Without tmux installed there is no session to
    open and no watcher to feed, so the bare interactive shell stands.
    """
    cmd = shell_command(prompt_file, done_path)
    shell = user_shell()
    if shutil.which("tmux") is None:
        return [shell, "-i", "-c", cmd]
    # One string, because that is the single shell-command argument `new-session`
    # takes; tmux hands it to `sh -c`, which splits it back into the argv above.
    return ["tmux", "new-session", f"{shlex.quote(shell)} -i -c {shlex.quote(cmd)}"]


def spawn_env() -> dict:
    """The environment a spawned agent gets: this process's, minus the variables
    that say we are *already* inside a tmux pane.

    Whoever launched the applet (or the mesh node) may have done so from one, and
    ``tmux new-session`` inherits ``$TMUX`` from there and refuses to nest — the
    window would open and close again on "sessions should be nested with care".
    Dropping them costs the child nothing: it is about to be in a pane of its own,
    which sets both to that pane's real values.
    """
    return {k: v for k, v in os.environ.items() if k not in ("TMUX", "TMUX_PANE")}


def spawn(prompt: str, preferred: SpawnTerminal | None, done_path: str | None = None) -> str:
    """Stage the prompt, open a new terminal window, run claude. Returns the
    prompt file path. Fully detached from the applet. ``done_path`` (optional)
    receives claude's exit code on completion — see :func:`shell_command`."""
    term = resolved(preferred)
    file = write_prompt(prompt)
    argv = [term.exec_name, *term.prefix, *agent_argv(file, done_path)]
    try:
        popen_detached(argv, env=spawn_env())
    except OSError as exc:
        raise SpawnError(f"failed to launch {term.title}: {exc}") from exc
    return file
